src/main/perl/RuleBasedEditor.pm

```diff
@@ -0,0 +1,1175 @@
+    config_rules: config rules corresponding to the file to build
+    config_options: configuration parameters used to build actual configuration
+    parser_options: a hash setting options to modify the behaviour of this function
+
+Supported entries for options hash:
+    always_rules_only: if true, apply only rules with ALWAYS condition (D: false)
+    remove_if_undef: if true, remove matching configuration line is rule condition is not met (D: false)
+
+Return value:
+    None or 1 in case of an internal error (missing argument)
+
+=cut
+
+sub _apply_rules
+{
+    my $function_name = "_apply_rules";
+    my ($self, $config_rules, $config_options, $parser_options) = @_;
+
+    unless ($config_rules) {
+        *$self->{LOG}->error("$function_name: 'config_rules' argument missing (internal error)");
+        return 1;
+    }
+    unless ($config_options) {
+        *$self->{LOG}->error("$function_name: 'config_options' argument missing (internal error)");
+        return 1;
+    }
+    unless (defined($parser_options)) {
+        *$self->{LOG}->debug(2, "$function_name: 'parser_options' undefined");
+        $parser_options = {};
+    }
+    if (defined($parser_options->{remove_if_undef})) {
+        *$self->{LOG}->debug(1, "$function_name: 'remove_if_undef' option set to " . $parser_options->{remove_if_undef});
+    } else {
+        *$self->{LOG}->debug(1, "$function_name: 'remove_if_undef' option not defined: assuming $LINE_OPT_DEF_REMOVE_IF_UNDEF");
+        $parser_options->{remove_if_undef} = $LINE_OPT_DEF_REMOVE_IF_UNDEF;
+    }
+
+
+    # Loop over all config rule entries, sorted by keyword alphabetical order.
+    # Config rules are stored in a hash whose key is the variable to write
+    # and whose value is the rule itself.
+    # If the variable name start with a '-', this means that the matching configuration
+    # line must be commented out unconditionally.
+    # Each rule format is '[condition->]attribute:option_set[,option_set,...];line_fmt' where
+    #     condition: either a role that must be enabled or ALWAYS if the rule must be applied
+    #                when 'always_rules_only' is true. A role is defined by an option set (see
+    #                Description at the beginning of this file, basically a sub-hash in the config)
+    #                and it is enabled if 'role_enabled' is true in the corresponding option set.
+    #     option_set and attribute: attribute in option set that must be substituted
+    #     line_fmt: the format to use when building the line
+    # An empty rule is valid and means that the keyword part must be
+    # written as is, using the line_fmt specified.
+
+    my $rule_id = 0;
+    foreach my $keyword (sort keys %$config_rules) {
+        my $rule = $config_rules->{$keyword};
+        $rule = '' unless defined($rule);
+        $rule_id++;
+
+        # Initialize parser_options for this rule according the default for this file
+        my $rule_parsing_options = {%{$parser_options}};
```

Comment:
remove the inner `{}`; also this is not a deep copy (but i have no clue what the values of parser_options are/can be)